// @copilotkit/react-core/dist/index.mjs
"use client";
import "./chunk-YJLRG5U3.mjs";
import {
  extract
} from "./chunk-ZM6HV22S.mjs";
import "./chunk-LUGEI4YQ.mjs";
import "./chunk-EFZPSZWO.mjs";
import {
  CopilotTask
} from "./chunk-7DFHFZJT.mjs";
import "./chunk-BWYAGPEF.mjs";
import {
  useCoAgent
} from "./chunk-QL23NJ57.mjs";
import {
  useCopilotAction
} from "./chunk-RYDEG77L.mjs";
import {
  useCopilotChat
} from "./chunk-MX2OYLRS.mjs";
import {
  useCopilotReadable
} from "./chunk-ANOG3W5S.mjs";
import {
  useMakeCopilotDocumentReadable
} from "./chunk-TRXKPXDE.mjs";
import "./chunk-JD7BAH7U.mjs";
import "./chunk-SPCZTZCY.mjs";
import {
  CopilotKit,
  defaultCopilotContextCategories
} from "./chunk-6YEMNWKE.mjs";
import "./chunk-XXR4QFAQ.mjs";
import "./chunk-5FHSUKQL.mjs";
import "./chunk-6U3UH3KO.mjs";
import "./chunk-YPSGKPDA.mjs";
import {
  CopilotMessagesContext,
  useCopilotMessagesContext
} from "./chunk-DCTJZ742.mjs";
import "./chunk-G2K6T2QU.mjs";
import {
  useCoAgentStateRender
} from "./chunk-CQ23Q555.mjs";
import {
  CopilotContext,
  useCopilotContext
} from "./chunk-2MPUQRAY.mjs";
import "./chunk-SKC7AJIV.mjs";
export {
  CopilotContext,
  CopilotKit,
  CopilotMessagesContext,
  CopilotTask,
  defaultCopilotContextCategories,
  extract,
  useCoAgent,
  useCoAgentStateRender,
  useCopilotAction,
  useCopilotChat,
  useCopilotContext,
  useCopilotMessagesContext,
  useCopilotReadable,
  useMakeCopilotDocumentReadable
};
